pkg/utils/kubernetes/networking/services.go

```diff
@@ -45,3 +48,92 @@ func WaitForServiceLoadBalancerAddress(ctx context.Context, c *kubernetes.Client
 		}
 	}
 }
+
+// WaitForServicePortConnected waits for a service provided by namespace/name
+// being able to connect by TCP. It creates a LoadBalancer with same selector and ports,
+// and try to establish TCP connection to the loadbalacer IP and port.
+func WaitForServicePortConnected(ctx context.Context, c *kubernetes.Clientset, namespace, name string, port int) error {
+	service, err := c.CoreV1().Services(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	lbServiceName := "templb-" + name
+	tempLoadBalancer := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      lbServiceName,
+			Labels: map[string]string{
+				"corresponding-service": name,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeLoadBalancer,
+			// copy the selector and ports of the service to check.
+			Selector: service.Spec.Selector,
+			Ports:    service.Spec.Ports,
+		},
+	}
+
+	// empty selector, we should create the endpoints separately.
+	if service.Spec.Selector == nil || len(service.Spec.Selector) == 0 {
+		endpoints, err := c.CoreV1().Endpoints(namespace).Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// copy the endpoints for the lb service, and set metadata.
+		tempEndPioints := endpoints.DeepCopy()
+		tempEndPioints.ObjectMeta = metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      lbServiceName,
+			Labels: map[string]string{
+				"corresponding-service": name,
+			},
+		}
+		_, err = c.CoreV1().Endpoints(namespace).Create(ctx, tempEndPioints, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			err := c.CoreV1().Endpoints(namespace).Delete(ctx, lbServiceName, metav1.DeleteOptions{})
+			if err != nil {
+				fmt.Printf("failed to delete endpoints %s/%s after testing, error %v",
+					namespace, lbServiceName, err)
+			}
+		}()
+	}
+
+	_, err = c.CoreV1().Services(namespace).Create(ctx, tempLoadBalancer, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err := c.CoreV1().Services(namespace).Delete(ctx, lbServiceName, metav1.DeleteOptions{})
+		if err != nil {
+			fmt.Printf("failed to delete service %s/%s after testing, error %v",
+				namespace, lbServiceName, err)
+		}
+	}()
+
+	ip, _, err := WaitForServiceLoadBalancerAddress(ctx, c, namespace, lbServiceName)
+	if err != nil {
+		return err
+	}
+
+	ticker := time.NewTicker(time.Second)
+	address := fmt.Sprintf("%s:%d", ip, port)
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context completed while waiting for %s:%d to be connected", ip, port)
+		case <-ticker.C:
+			// TODO: add a timeout for connecting to the address and make it configurable?
```

Comment:
Yes, we should use a timeout for this. You can find some examples of this in the [KIC TCPRoute Tests](https://github.com/Kong/kubernetes-ingress-controller/blob/main/test/integration/tcproute_test.go)

Reply:
updated, added a `dialTimeout` parameter.